src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4366,49 +4364,45 @@ public String descriptorString() {
      *           may be removed in a future release, or upgraded to permanent
      *           features of the Java language.}
      *
-     * Returns an array containing {@code ClassDesc} objects representing all the
+     * Returns an array containing {@code Class} objects representing all the
      * direct subclasses or direct implementation classes permitted to extend or
      * implement this class or interface if it is sealed. The order of such elements
      * is unspecified. If this {@code Class} object represents a primitive type,
      * {@code void}, an array type, or a class or interface that is not sealed,
      * an empty array is returned.
      *
-     * @return an array of class descriptors of all the permitted subclasses of this class or interface
+     * @return an array of class objects of all the permitted subclasses of this class or interface
      *
      * @jls 8.1 Class Declarations
      * @jls 9.1 Interface Declarations
      * @since 15
      */
     @jdk.internal.PreviewFeature(feature=jdk.internal.PreviewFeature.Feature.SEALED_CLASSES, essentialAPI=false)
-    public ClassDesc[] permittedSubclasses() {
-        String[] subclassNames;
-        if (isArray() || isPrimitive() || (subclassNames = getPermittedSubclasses0()).length == 0) {
-            return EMPTY_CLASS_DESC_ARRAY;
-        }
-        ClassDesc[] constants = new ClassDesc[subclassNames.length];
-        int i = 0;
-        for (String subclassName : subclassNames) {
-            try {
-                constants[i++] = ClassDesc.of(subclassName.replace('/', '.'));
-            } catch (IllegalArgumentException iae) {
-                throw new InternalError("Invalid type in permitted subclasses information: " + subclassName, iae);
-            }
+    public Class<?>[] getPermittedSubclasses() {
```

Comment:
What happens if a permitted subclass is not found?   I see that `getPermittedSubclasses0` ignores the entry if the class is not found.    Should that be specified?

Have you considered whether security package access is needed (now that this method returns `Class` objects the caller may not have access to)? This needs to be discussed with the security team.    If someone gets a hold of a sealed class (e.g. `obj.getClass()`), this method could leak other `Class` objects.